mempool/src/pool.rs

```diff
@@ -0,0 +1,815 @@
+                || Ok(()),
+                |outpoint| {
+                    Err(TxValidationError::OutPointNotFound {
+                        outpoint: outpoint.clone(),
+                        tx: tx.clone(),
+                    })
+                },
+            )
+    }
+
+    // This function returns true even if the outpoint is marked as Spent because of another
+    // transaction that has been accepted to the mempool
+    fn outpoint_available(&self, outpoint: &OutPoint) -> bool {
+        self.store.contains_outpoint(outpoint) || self.chain_state.contains_outpoint(outpoint)
+    }
+
+    fn is_replaceable(&self, tx: Arc<Transaction>) -> bool {
+        tx.is_replaceable()
+            || self.store.ancestors(&tx.into()).iter().any(|entry| entry.tx.is_replaceable())
+    }
+
+    fn validate_transaction(&self, tx: &Transaction) -> Result<(), TxValidationError> {
+        if tx.get_inputs().is_empty() {
+            return Err(TxValidationError::NoInputs);
+        }
+
+        if tx.get_outputs().is_empty() {
+            return Err(TxValidationError::NoOutputs);
+        }
+
+        if tx.is_coinbase() {
+            return Err(TxValidationError::LooseCoinbase);
+        }
+
+        // TODO consier a MAX_MONEY check reminiscent of bitcoin's
+        // TODO consider rejecting non-standard transactions
+        if has_duplicate_entry(tx.get_inputs()) {
+            return Err(TxValidationError::DuplicateInputs);
+        }
+
+        if tx.encoded_size() > MAX_BLOCK_SIZE {
+            return Err(TxValidationError::ExceedsMaxBlockSize);
+        }
+
+        if self.contains_transaction(&tx.get_id()) {
+            return Err(TxValidationError::TransactionAlreadyInMempool);
+        }
+
+        tx.get_inputs()
+            .iter()
+            .filter_map(|input| self.store.find_conflicting_tx(input.get_outpoint()))
+            .all(|tx| self.is_replaceable(tx))
+            .then(|| ())
+            .ok_or(TxValidationError::ConflictWithIrreplaceableTransaction)?;
+
+        self.verify_inputs_available(tx)?;
+
+        //
+        // TODO Understand why we need  two size checks, one in CheckTransaction, and one in
+        // TODO Do we want to differentiate between standard and nonstandard transaction
+        // TODO do we really want lock time to be u32? I think it should be an enum
+        // TODO check finality, both tx wise and output wise
+        //
+
+        Ok(())
+    }
+}
+
+impl<C: ChainState + Debug> Mempool<C> for MempoolImpl<C> {
+    // TODO need to discuss parameters
+    fn create(chain_state: C) -> Self {
+        Self {
+            store: MempoolStore::new(),
+            chain_state,
+        }
+    }
+    // TODO what are the inputs for this?
+    fn new_tip_set(&mut self) -> Result<(), MempoolError> {
+        Err(MempoolError::GenericError)
+    }
+    //
+
+    fn add_transaction(&mut self, tx: Transaction) -> Result<(), MempoolError> {
+        if self.store.txs_by_fee.len() >= MEMPOOL_MAX_TXS {
```

Comment:
Keep in mind that dropping a transaction just for its fee can leave the mempool in an invalid state, because that transaction could be input for another transaction in the mempool. It's a consistently held invariant that all transactions in the mempool should be valid and directly usable. Otherwise, transactions go to orphans.

Reply:
Yup. Dropping a transaction means dropping all in-mempool, descendants, as far I understand. So if anything a transaction should be dropped based on a low "score", where score is some number which takes into account the fees of its descendants